components/calendar/src/islamic.rs

```diff
@@ -308,6 +305,267 @@ impl DateTime<IslamicObservational> {
+    ///
+    /// ```rust
+    /// use icu::calendar::Date;
+    ///
+    /// let date_islamic = Date::try_new_ummalqura_date(1392, 4, 25)
+    ///     .expect("Failed to initialize Islamic Date instance.");
+    ///
+    /// assert_eq!(date_islamic.year().number, 1392);
+    /// assert_eq!(date_islamic.month().ordinal, 4);
+    /// assert_eq!(date_islamic.day_of_month().0, 25);
+    /// ```
+    pub fn try_new_ummalqura_date(
+        year: i32,
+        month: u8,
+        day: u8,
+    ) -> Result<Date<UmmalQura>, CalendarError> {
+        ArithmeticDate::new_from_lunar_ordinals(year, month, day)
+            .map(UmmalQuraDateInner)
+            .map(|inner| Date::from_raw(inner, UmmalQura))
+    }
+}
+
+impl DateTime<UmmalQura> {
+    /// Construct a new UmmalQura datetime from integers.
+    ///
+    /// ```rust
+    /// use icu::calendar::DateTime;
+    ///
+    /// let datetime_islamic = DateTime::try_new_ummalqura_datetime(474, 10, 11, 13, 1, 0)
+    ///     .expect("Failed to initialize Islamic DateTime instance.");
+    ///
+    /// assert_eq!(datetime_islamic.date.year().number, 474);
+    /// assert_eq!(datetime_islamic.date.month().ordinal, 10);
+    /// assert_eq!(datetime_islamic.date.day_of_month().0, 11);
+    /// assert_eq!(datetime_islamic.time.hour.number(), 13);
+    /// assert_eq!(datetime_islamic.time.minute.number(), 1);
+    /// assert_eq!(datetime_islamic.time.second.number(), 0);
+    /// ```
+    pub fn try_new_ummalqura_datetime(
+        year: i32,
+        month: u8,
+        day: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+    ) -> Result<DateTime<UmmalQura>, CalendarError> {
+        Ok(DateTime {
+            date: Date::try_new_ummalqura_date(year, month, day)?,
+            time: types::Time::try_new(hour, minute, second, 0)?,
+        })
+    }
+}
+
+impl UmmalQura {
+    /// Constructs a new UmmalQura Islamic Calendar
+    pub fn new() -> Self {
+        Self
+    }
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6957
+    fn saudi_criterion(date: RataDie) -> Option<bool> {
+        let sunset = Astronomical::sunset((date - 1).as_moment(), MECCA)?;
+        let tee = Location::universal_from_standard(sunset, MECCA);
+        let phase = Astronomical::lunar_phase(tee);
+        let moonlag = Astronomical::moonlag((date - 1).as_moment(), MECCA)?;
+
+        Some(phase > 0.0 && phase < 90.0 && moonlag > 0.0)
+    }
+    pub(crate) fn adjusted_saudi_criterion(date: RataDie) -> bool {
+        if let Some(x) = Self::saudi_criterion(date) {
+            x
+        } else {
+            false
+        }
+    }
+
+    fn saudi_new_month_on_or_before(date: RataDie) -> RataDie {
+        let moon =
+            libm::floor((Astronomical::lunar_phase_at_or_before(0.0, date.as_moment())).inner());
+        let age = date.to_f64_date() - moon;
+        let tau = if age <= 3.0 && !Self::adjusted_saudi_criterion(date) {
```

Comment:
`tau` is an approximation, our theory is that `3.0` is simply a number that is not too small to avoid off-by-one-day errors, but is less than 14 and ensures that we are on the right "side" of the month

@eggrobin may wish to try and understand this better

Reply:
note to self: moonrise happens later every day (by approx 0.5-2hrs)

Reply:
Okay, trying to understand the number 3 here. TLDR I think it's justchosen such that it doesn't matter.

Firstly, note that the saudi criterion is calculated with sunset during the _previous_ day. To keep things straight, when I speak of the criterion holding _for_ a day I mean wrt sunset the previous day, and when I speak of it holding _on_ a day sunset I mean actually on that ISO day (midnight - midnight). 

This is partly the source of a lot of confusion here: you have multiple notions of "day" vying for space.

Also note that `moon` is an integer value (since we `floor()` it), and `date.as_moment()` normalizes the date to midnight at the beginning of the day.

For the first day of the month; the saudi criterion will be valid (by definition); and the saudi criterion will be valid in its near future as well. Because the first day of the month is near the new moon, moonset cannot be _much_ later than sunset so there is no chance of moonset spilling over into the next day's calculation.


Going to first work through cases.

Let's take the case that the new moon happens on Tuesday night, and today is Wednesday. The saudi criterion will not hold on Tuesday sunset (the moon was not a waxing crescent when it set), i.e. it does not hold for Wednesday. It will hold Wednesday sunset, since after the new moon the sun must set first. `age < 3` (`age` is `1`, since `moon` is Tuesday) and `saudi(today) == false`. Thursday is the new month; we are in the previous month. 

Let's take the case that the new moon happens on Monday night, and today is Wednesday. The saudi criterion will not hold for Tuesday (same reason as before). It will hold for Wednesday, since the moon _must_ set after Tuesday sunset. `age` is now `2` but `saudi(today) == true`. Today is the new month. The same situation occurs for the new moon on Tuesday morning, with `age` as `1`.



When the saudi criterion is false for today, only the following cases may cause it:

 - The moon is currently not a waxing crescent (`age ⪆ 7.5`)
 - The new moon happened today after sunset. This will still be `age ≈ 30`, this is equivalent to "the moon is not a waxing crescent for any of the times that matter.
 -  The moon is a waxing crescent, but the new moon happened last night (and the moon was a waning crescent last sunset). `age` is 1 here.
 - The new moon happened exactly at midnight this morning. This is `age == 0`, however, the moon will not have been a waxing crescent yesterday sunset.

There are no other cases: any time shortly after the new moon; the moon must set after the sun.

The latter two cases are the ones where the value of `moon` will be wrong, since, yes, the last new moon was "yesterday" (or today morning-midnight), but it was not during the part of yesterday that matters, so we need to go back a month.


So _really_ all that matters here is `age <= 1 && !saudi(today)`. `!saudi(today)` is impossible with `age == 2`. The number `3` appears chosen as a number far enough from 1 or 7 such that one does not have to think too deeply about it.